x-pack/packages/kbn-elastic-assistant/impl/assistant/use_conversation/index.tsx

```diff
@@ -51,299 +48,264 @@ interface CreateConversationProps {
+          http,
+          conversationId,
+          messages,
+        });
+      }
       return messages;
     },
-    [setConversations]
+    [http]
   );
 
   /**
    * Updates the last message of conversation[] for a given conversationId with provided content
    */
   const amendMessage = useCallback(
-    ({ conversationId, content }: AmendMessageProps) => {
-      setConversations((prev: Record<string, Conversation>) => {
-        const prevConversation: Conversation | undefined = prev[conversationId];
-
-        if (prevConversation != null) {
-          const { messages, ...rest } = prevConversation;
-          const message = messages[messages.length - 1];
-          const updatedMessages = message
-            ? [...messages.slice(0, -1), { ...message, content }]
-            : [...messages];
-          const newConversation = {
-            ...rest,
-            messages: updatedMessages,
-          };
-          return {
-            ...prev,
-            [conversationId]: newConversation,
-          };
-        } else {
-          return prev;
-        }
-      });
+    async ({ conversationId, content }: AmendMessageProps) => {
+      const prevConversation = await getConversationById({ http, id: conversationId });
+      if (isHttpFetchError(prevConversation)) {
+        return;
+      }
+      if (prevConversation != null) {
+        const { messages } = prevConversation;
+        const message = messages[messages.length - 1];
+        const updatedMessages = message ? [{ ...message, content }] : [];
+
+        await appendConversationMessagesApi({
+          http,
+          conversationId,
+          messages: updatedMessages,
+        });
+      }
     },
-    [setConversations]
+    [http]
   );
 
   /**
    * Append a message to the conversation[] for a given conversationId
    */
   const appendMessage = useCallback(
-    ({ conversationId, message }: AppendMessageProps): Message[] => {
+    async ({ conversationId, message }: AppendMessageProps): Promise<Message[] | undefined> => {
       assistantTelemetry?.reportAssistantMessageSent({
         conversationId,
         role: message.role,
         isEnabledKnowledgeBase,
         isEnabledRAGAlerts,
       });
-      let messages: Message[] = [];
-      setConversations((prev: Record<string, Conversation>) => {
-        const prevConversation: Conversation | undefined = prev[conversationId];
 
-        if (prevConversation != null) {
-          messages = [...prevConversation.messages, message];
-          const newConversation = {
-            ...prevConversation,
-            messages,
-          };
-          return {
-            ...prev,
-            [conversationId]: newConversation,
-          };
-        } else {
-          return prev;
-        }
+      const res = await appendConversationMessagesApi({
+        http,
+        conversationId,
+        messages: [message],
       });
-      return messages;
+      if (isHttpFetchError(res)) {
+        return;
```

Comment:
Consider displaying an error toast indicating the latest message won't be persisted to the conversation